components/function-controller/internal/controllers/serverless/deployment.go

```diff
@@ -90,18 +82,21 @@ func (r *FunctionReconciler) equalDeployments(existing appsv1.Deployment, expect
 		equalResources(existing.Spec.Template.Spec.Containers[0].Resources, expected.Spec.Template.Spec.Containers[0].Resources)
 }
 
-func (r *FunctionReconciler) updateDeploymentStatus(ctx context.Context, log logr.Logger, instance *serverlessv1alpha1.Function, deployment appsv1.Deployment) (ctrl.Result, error) {
+func (r *FunctionReconciler) updateDeploymentStatus(ctx context.Context, log logr.Logger, instance *serverlessv1alpha1.Function, deployment appsv1.Deployment, runningStatus corev1.ConditionStatus) (ctrl.Result, error) {
 	switch {
-	case r.isDeploymentReady(deployment):
+	// this step is both in onDeploymentChange and as last step in reconcile
+	// it's checked here in onDeploymentChange to prevent nasty data races, where somehow deployment becomes ready before we
+	// trigger next reconcile loop, in which we should create svc
+	case r.isDeploymentInCondition(deployment, appsv1.DeploymentAvailable):
 		log.Info(fmt.Sprintf("Deployment %s is ready", deployment.GetName()))
 		return r.updateStatus(ctx, ctrl.Result{}, instance, serverlessv1alpha1.Condition{
 			Type:               serverlessv1alpha1.ConditionRunning,
-			Status:             corev1.ConditionUnknown,
+			Status:             runningStatus,
 			LastTransitionTime: metav1.Now(),
 			Reason:             serverlessv1alpha1.ConditionReasonDeploymentReady,
```

Comment:
This is only a question/sugestion. In this case we should point to last reason - for example, if hpa will be the last reconciled item, then we should point to last reason from Function (probably pointed to HPA), not change to `ConditionReasonDeploymentReady`. Of course, if deployment has change, then we must force update status to appropriate from Deployment, but when we change status from main reconcile function (default case from last `switch`), then we should retrieve reason from Function instance.

Reply:
But in this reconcile deployment IS the last resource reconciled, as it goes like this:
- ...
- create deployment
- create svc
- create hpa 
- wait for deployment to be ready, if it is set proper status

If we wanted hpa to be last reconciled item, then we'd need to add more states, which would make our external API look really confusing (then we'd have to ditch `Running` state and add something like `HPAcreated` and others, which is confusing)